blocks/blocks.php

```diff
@@ -47,11 +26,97 @@ function parse_blocks( $post_model ) {
 	// based on "edit_posts" cap. Since we want to serve blocks even to logged-out
 	// users -- and because we are parsing this content before returning it --
 	// we'll bypass the model and access the raw content directly.
-	$post = get_post( $post_model->ID );
+	$post = \get_post( $post_model->ID );
 
 	$is_gutenberg = \has_blocks( $post->post_content );
-	$blocks       = \parse_blocks( $post->post_content );
+	$raw_blocks   = \parse_blocks( $post->post_content );
+
+	return [
+		'blocks'      => process_content_blocks( $raw_blocks ),
+		'isGutenberg' => $is_gutenberg,
+		'version'     => $version,
+	];
+}
+
+/**
+ * Map the Gutenberg block attributes to the shape of BlockAttribute.
+ *
+ * @param  array $block Content block.
+ * @return array
+ */
+function get_content_block_attributes( $block ) {
+	if ( 'core/image' === $block['blockName'] ) {
+		$attachment_metadata = \wp_get_attachment_metadata( $block['attrs']['id'] );
+
+		$block['attrs']['src']            = \wp_get_attachment_url( $block['attrs']['id'] );
+		$block['attrs']['originalHeight'] = $attachment_metadata['height'];
+		$block['attrs']['originalWidth']  = $attachment_metadata['width'];
+		$block['attrs']['srcset']         = \wp_get_attachment_image_srcset( $block['attrs']['id'] );
+		$block['attrs']['alt']            = trim( strip_tags( \get_post_meta( $block['attrs']['id'], '_wp_attachment_image_alt', true ) ) );
+
+		// If width and height attributes aren't exposed, add the default ones
+		if ( ! isset( $block['attrs']['height'] ) ) {
+			$block['attrs']['height'] = $attachment_metadata['height'];
+		}
+
+		if ( ! isset( $block['attrs']['width'] ) ) {
+			$block['attrs']['width'] = $attachment_metadata['width'];
+		}
+	}
+
+	return array_map(
+		function ( $key ) use ( $block ) {
+			return [
+				'name'  => $key,
+				'value' => $block['attrs'][ $key ],
+			];
+		},
+		array_keys( $block['attrs'] )
+	);
+}
+
+/**
+ * Strip wrapping tags from the content and set as a property on the block. This
+ * allows the front-end implementor to delegate tag creation to a component.
+ *
+ * @param  string $html Inner HTML of block.
+ * @return array
+ */
+function get_content_block_html( $html ) {
+	$html = trim( $html );
+
+	preg_match( '#^<([A-z][A-z0-9]*)\b([^>])*>(.*?)</\1>$#', $html, $matches );
+
+	if ( isset( $matches[1] ) ) {
+		return [
+			'innerHTML' => $matches[3],
+			'outerHTML' => $html,
+			'tagName'   => $matches[1],
+		];
+	}
```

Comment:
I'm curious about the decision here to change the properties from what `parse_blocks()` implements out of the box? As a WP developer, I would expect the `innerHTML` property of a block to be equivalent to what the block parser returns. But here, it looks like GraphQL would return just the markup found inside the wrapping element and return the equivalent as `outerHTML`.

I like the intent of trying to return the wrapping tag name and inner markup separately, but worry that this would be confusing as it's inconsistent with what other implementations of block data would call `innerHTML`.

Additionally, `parse_blocks()` returns an `innerContent` property which can be useful to include in responses so a custom component can reconstruct the shape of a block that mixes `innerBlocks` and `innerHTML`. 

See: https://github.com/WordPress/WordPress/blob/5.8-branch/wp-includes/class-wp-block-parser.php#L56-L68

Reply:
I added `innerContent` in https://github.com/Automattic/vip-decoupled-bundle/pull/38/commits/2eefb78b55116a8f99ee54ddc0decd64ae248037. Great suggestion.

> As a WP developer, I would expect the innerHTML property of a block to be equivalent to what the block parser returns. But here, it looks like GraphQL would return just the markup found inside the wrapping element and return the equivalent as outerHTML.

@joemcgill The reason we made this choice is that it matches how this data is often consumed by component-driven front-ends. Keeping the wrapping tag in the `innerHTML` string makes it impossible to do something like this:

```
function Paragraph ( props ) {
	return <p dangerouslySetInnerHTML={ { __html: props.block.innerHTML }} />;
}
```

I know your point is that it's confusing to call it `innerHTML` when it differs from what the Gutenberg parser calls `innerHTML`, but this aligns with how React and other frameworks think about "innerHTML." Of course, we could call it something else here like `componentHtml`, but we'd lose that alignment. I also don't know how many people are familiar enough with the output of `parse_blocks` for this to be genuinely confusing. 

But you raise a good enough point that I am a bit conflicted. Any further thoughts?

Reply:
> I also don't know how many people are familiar enough with the output of parse_blocks for this to be genuinely confusing.

I'm equally conflicted about this, because if I were _not_ expecting the GraphQL response to mimic the data for blocks as it is modeled inside WordPress, I would want `innerHTML` and `outerHTML` to work exactly as you are proposing here. However, I think WordPress is assuming that `innerHTML` references all of the HTML that is contained inside the block delimiters, and not just what is inside the  outermost wrapping element. For that matter, I _think_ it's technically possible with v2 of the blocks API for there to be multiple elements inside a block's `innerHTML`, like this for example:

```html
<!-- example:my-block -->
<div>
  <p>Some text</p>
</div>
<div>
  <p>Some other text</p>
</div>
<!-- /example:my-block -->
```

If that's true (I've not verified), then returning all of the content between the delimiters as `innerHTML` makes the most sense to me.

The new `WP_REST_Block_Renderer_Controller` that will be included in WP 5.9 takes the approach of modeling block content based on the `parse_blocks` data ([see reference](https://github.com/WordPress/wordpress-develop/blob/aa99d7e7a34a5f460ddb85f2e009273408fc0585/src/wp-includes/rest-api/endpoints/class-wp-rest-block-renderer-controller.php#L181-L186)), so there's a case to be made that the block parser modeling is the defacto standard for how developers will interact with block data in various contexts.

Reply:
Hey @joemcgill! Thanks for the conversation and pointers here. The idea that the output of `parse_blocks` is the de facto schema for block data under the hood is pretty exciting and convincing enough for me. I've made a few changes in https://github.com/Automattic/vip-decoupled-bundle/pull/38/commits/f560e0356ed267c6cd569e642af37a216472a239 as a result:

1. Deprecate `outerHTML` (still available, just marked as deprecated).
2. By default, do not attempt to strip the wrapping tag from `innerHTML`.
3. Introduce a boolean `removeWrappingTag` field directive to `innerHTML` (default: `false`) that allows queries to request the removal of the wrapping tag if they find it useful:

```
fragment ContentBlockParts {
  name
  tagName
  innerHTML(removeWrappingTag: true)
  attributes {
    name
    value
  }
}
```

If requested and it's not possible to remove the wrapping tag (i.e., because of the multiple element scenario you posed), then the field will return null.

Interested if this feels more aligned to you.

Reply:
Thanks @chriszarate. This new approach does seem more aligned to me. 

The only problem I'm seeing now when testing is that the `innerHTML` and `outerHTML` are both returning null right now. Here's the query I'm testing in the GraphiQL IDE and the result I'm seeing.

**Query:**
```json
{
  pages {
    nodes {
      contentBlocks {
        blocks {
          attributes {
            name
            value
          }
          innerContent
          innerHTML
          name
          outerHTML
          tagName
        }
      }
    }
  }
}
``` 
**Result:**
```
{
  "data": {
    "pages": {
      "nodes": [
        {
          "contentBlocks": {
            "blocks": [
              {
                "attributes": [],
                "innerContent": [
                  "\n<p>This is an example page. It's different from a blog post because it will stay in one place and will show up in your site navigation (in most themes). Most people start with an About page that introduces them to potential site visitors. It might say something like this:</p>\n"
                ],
                "innerHTML": null,
                "name": "core/paragraph",
                "outerHTML": null,
                "tagName": "p"
              }
            ]
          }
        }
      ]
    }
  }
}
```

If I update the query to pass `innerHTML(removeWrappingTag: true)` then I get the expected result of the innerHTML with the wrapping `p` tags stripped. Explicitly passing false also returns `null`.

Reply:
@joemcgill Sorry about that, pushed up a one-liner (https://github.com/Automattic/vip-decoupled-bundle/pull/38/commits/1b49d4092c88a0ee49e03870dce53a2c09cd3d0b) to fix that. Will also be adding tests, soon. :)

Reply:
Yup. That did it 👍🏻. This looks really great. Thanks for all the effort you've put into it.